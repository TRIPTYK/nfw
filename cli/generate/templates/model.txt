import { Entity, PrimaryGeneratedColumn, Column, ManyToOne } from "typeorm";
import { DateUtils } from "typeorm/util/DateUtils";
import { {{ENTITY_CAPITALIZE}}Serializer } from "./../serializers/{{ENTITY_LOWERCASE}}.serializer";

@Entity()
export class {{ENTITY_CAPITALIZE}} {

  /**
   * @param payload Object data to assign
   */
  constructor(payload: Object ) { Object.assign(this, payload); }

  @PrimaryGeneratedColumn()
  id: Number;

  @Column({
    type: Date,
    default: DateUtils.mixedDateToDateString( new Date() )
  })
  createdAt;

  @Column({
    type: Date,
    default: null
  })
  updatedAt;

  @Column({
    type: Date,
    default: null
  })
  deletedAt;

  public whitelist() {
    return new {{ENTITY_CAPITALIZE}}Serializer().serializer.serialize(this);
  } 
}