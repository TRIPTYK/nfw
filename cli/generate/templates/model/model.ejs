import { Entity, PrimaryGeneratedColumn, Column, ManyToOne , OneToOne , JoinColumn , OneToMany , ManyToMany , CreateDateColumn , UpdateDateColumn } from "typeorm";
import { DateUtils } from "typeorm/util/DateUtils";
import { <%- entityCapitalize %>Serializer } from "./../serializers/<%- entityLowercase %>.serializer";
<% imports.forEach(imprt => {%>
<%- imprt -%>
<% }); %>

@Entity()
export class <%- entityCapitalize %> {

  /**
   * @param payload Object data to assign
   */
  constructor(payload: Object ) { Object.assign(this, payload); }

  @PrimaryGeneratedColumn()
  id: Number;

<% entities.forEach(entity => {%>
  <%- include('/model/_column',{entity}) -%>
<%});-%>

<% foreignKeys.forEach(foreignKey => {%>
  <%- include('/model/_foreignKey',{foreignKey}) -%>
<%});-%>

<% if (createUpdate && createUpdate.createAt) { %> @CreateDateColumn() <% } -%>
<% if (createUpdate && createUpdate.updateAt) { %> @UpdateDateColumn() <% } -%>

  public whitelist() {
    return new <%- entityCapitalize %>Serializer().serializer.serialize(this);
  }
}
