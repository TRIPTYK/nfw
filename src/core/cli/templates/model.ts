import project = require("../utils/project");
import { SourceFile } from "ts-morph";
import { GeneratorParameters } from "../interfaces/generator.interface";
import  * as pluralize from "pluralize";

/**
 *
 * @param path
 * @param className
 * @param {array} entities
 * @return {SourceFile}
 */
export default function createModelTemplate({fileTemplateInfo,classPrefixName,modelName,filePrefixName}: GeneratorParameters): SourceFile {

    const file = project.createSourceFile(`${fileTemplateInfo.path}/${fileTemplateInfo.name}`,null,{
        overwrite : true
    });

    const interfaceNameForModel = `${classPrefixName}Interface`;

    file.addInterface({
        name : interfaceNameForModel
    }).setIsExported(true);

    file.addImportDeclaration({
        moduleSpecifier : `../validations/${filePrefixName}.validation`,
        defaultImport: `* as ${classPrefixName}Validator`
    })

    const modelClass = file.addClass({
        name: classPrefixName
    });

    modelClass.setExtends(`JsonApiModel<${classPrefixName}>`);
    modelClass.addImplements(interfaceNameForModel);
    modelClass.addDecorator({name : "JsonApiEntity",arguments : [`"${pluralize(modelName)}"`,(writer) => {
        writer.block(() => {
            writer.setIndentationLevel(1);
            writer.writeLine(`serializer: ${classPrefixName}Serializer,`);
            writer.writeLine(`repository: ${classPrefixName}Repository,`);
            writer.writeLine(`validator: ${classPrefixName}Validator`);
        });
    }
    ]}).setIsDecoratorFactory(true);
    modelClass.setIsExported(true);

    return file;
};
