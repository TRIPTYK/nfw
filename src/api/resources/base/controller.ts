import type { EntityData, RequiredEntityData } from '@mikro-orm/core';
import { wrap } from '@mikro-orm/core';
import type { JsonApiQuery, ResourceSerializer } from '@triptyk/nfw-resources';
import { canOrFail } from 'app/api/utils/can-or-fail.js';
import type { UserModel } from 'app/database/models/user.model.js';
import type { ResourceAuthorizer } from './authorizer.js';
import type { ResourceService } from './service.js';

export interface JsonApiContext<R extends Record<string, unknown>, T> {
    service: ResourceService<T>,
    authorizer: ResourceAuthorizer<T>,
    serializer: ResourceSerializer<R>,
}

export async function jsonApiGetFunction<T, R extends Record<string, unknown>> (this: JsonApiContext<R, T>, id: string, query: JsonApiQuery, currentUser: UserModel) {
  const resourceModel = await this.service.getOneOrFail(id, query);
  await canOrFail(this.authorizer, currentUser, 'read', resourceModel);
  return this.serializer.serializeOne(wrap(resourceModel).toJSON() as never, query);
}

export async function jsonApiFindAllFunction <T, R extends Record<string, unknown>> (this: JsonApiContext<R, T>, query: JsonApiQuery, currentUser: UserModel) {
  const [resourcesModel, count] = await this.service.getAll(query);
  await canOrFail(this.authorizer, currentUser, 'read', resourcesModel);
  return this.serializer.serializeMany(resourcesModel.map((d) => wrap(d).toJSON()) as never, query, query.page ? { ...query.page, total: count } : undefined);
}

export async function jsonApiCreateFunction<T, R extends Record<string, unknown>> (this: JsonApiContext<R, T>, currentUser: UserModel, body: RequiredEntityData<T>) {
  await canOrFail(this.authorizer, currentUser, 'create', body);
  const user = await this.service.create(body);
  return this.serializer.serializeOne(user as never, {});
}

export async function jsonApiUpdateFunction<T, R extends Record<string, unknown>> (this: JsonApiContext<R, T>, currentUser: UserModel, body: EntityData<T>, id: string) {
  await canOrFail(this.authorizer, currentUser, 'update', body);
  const user = await this.service.update(id, body);
  return this.serializer.serializeOne(user as never, {});
}

export async function jsonApiDeleteFunction<T, R extends Record<string, unknown>> (this: JsonApiContext<R, T>, id: string, currentUser: UserModel) {
  const user = await this.service.getOneOrFail(id, {});
  await canOrFail(this.authorizer, currentUser, 'delete', user);
  await this.service.delete(id);
  return null;
}
