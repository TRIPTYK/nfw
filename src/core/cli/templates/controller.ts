import project = require("../utils/project");
import { GeneratorParameters } from "../interfaces/generator.interface";

/**
 *
 * @param path
 * @param className
 * @param options
 * @param classPrefixName
 */
export default function createControllerTemplate({fileTemplateInfo,classPrefixName,filePrefixName}: GeneratorParameters) {
    const file = project.createSourceFile(`${fileTemplateInfo.path}/${fileTemplateInfo.name}`, null, {
        overwrite: true
    });


    file.addImportDeclaration({
        namedImports: [classPrefixName],
        moduleSpecifier: `../models/${filePrefixName}.model`,
    })

    const controllerClass = file.addClass({
        name: `${classPrefixName}Controller`
    });

    controllerClass.setIsDefaultExport(true);

    controllerClass.addDecorator({
        name: "JsonApiController",
        arguments: [`${classPrefixName}`]
    }).setIsDecoratorFactory(true);
    controllerClass.addDecorator({
        name: "singleton"
    }).setIsDecoratorFactory(true);
    controllerClass.addDecorator({
        name: "autoInjectable"
    }).setIsDecoratorFactory(true);

    controllerClass.setExtends(`BaseJsonApiController<${classPrefixName}>`);

    return file;
};
