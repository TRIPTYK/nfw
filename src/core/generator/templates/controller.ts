import { GeneratorParameters } from "../interfaces/generator.interface";
import project from "../utils/project";

/**
 *
 * @param path
 * @param className
 * @param options
 * @param classPrefixName
 */
export default function createControllerTemplate({
    fileTemplateInfo,
    classPrefixName,
    filePrefixName
}: GeneratorParameters) {
    const file = project.createSourceFile(
        `${fileTemplateInfo.path}/${fileTemplateInfo.name}`,
        null,
        {
            overwrite: true
        }
    );

    file.addImportDeclaration({
        namedImports: [classPrefixName],
        moduleSpecifier: `../models/${filePrefixName}.model`
    });

    const controllerClass = file.addClass({
        name: `${classPrefixName}Controller`
    });

    controllerClass.setIsDefaultExport(true);

    controllerClass
        .addDecorator({
            name: "JsonApiController",
            arguments: [`${classPrefixName}`]
        })
        .setIsDecoratorFactory(true);
    controllerClass
        .addDecorator({
            name: "singleton"
        })
        .setIsDecoratorFactory(true);
    controllerClass
        .addDecorator({
            name: "autoInjectable"
        })
        .setIsDecoratorFactory(true);

    controllerClass.setExtends(`BaseJsonApiController<${classPrefixName}>`);

    return file;
}
