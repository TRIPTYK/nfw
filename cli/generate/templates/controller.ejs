import * as HttpStatus from "http-status";

import { Request, Response } from "express";
import { <%- entityCapitalize %> } from "./../models/<%- entityLowercase %>.model";
import { <%- entityCapitalize %>Repository } from "./../repositories/<%- entityLowercase %>.repository";
import { getRepository, getCustomRepository } from "typeorm";
import { BaseController } from "./base.controller";
const promisify = require('util')

/**
 *
 */
export class <%- entityCapitalize %>Controller extends BaseController {

  /** */
  constructor() { super(); }

<% if (options.read) { %>
  /**
   * Get serialized <%- entityLowercase %>
   *
   * @param req Request
   * @param res Response
   *
   * @public
   */
   public async get(req: Request, res : Response) {
    const repository = getRepository(<%- entityCapitalize %>);
    const <%- entityLowercase %> = await repository.findOneOrFail(req.params.<%- entityLowercase %>Id);
    res.json( <%- entityLowercase %>.whitelist() );
  }
<% } %>

<% if (options.create) { %>
  /**
   * Create new <%- entityLowercase %>
   *
   * @param req Request
   * @param res Response
   * @param next Function
   *
   * @public
   */
  public async create (req: Request, res : Response, next: Function) {
    try {
      const repository = getRepository(<%- entityCapitalize %>);
      const <%- entityLowercase %> = new <%- entityCapitalize %>(req.body);
      const saved<%- entityCapitalize %> = await repository.save(<%- entityLowercase %>);
      res.status( HttpStatus.CREATED );
      res.json( saved<%- entityCapitalize %>.whitelist() );
    }
    catch (e) { next( e.message ); }
  }
<% } %>

<% if (options.update) { %>
  /**
   * Update existing <%- entityLowercase %>
   *
   * @param req Request
   * @param res Response
   * @param next Function
   *
   * @public
   */
  public async update (req: Request, res : Response, next: Function) {

    try {
      const repository = getRepository(<%- entityCapitalize %>);
      const <%- entityLowercase %> = await repository.findOne(req.params.<%- entityLowercase %>Id);
      repository.merge(<%- entityLowercase %>, req.body);
      repository.save(<%- entityLowercase %>);
      res.json( <%- entityLowercase %>.whitelist() );
    }
    catch(e) { next( e.message ); }

  };
<% } %>

<% if (options.read) {%>
  /**
   * Get <%- entityLowercase %> list
   *
   * @param req Request
   * @param res Response
   * @param next Function
   *
   * @public
   */
  public async list (req: Request, res : Response, next: Function) {

    try {
      const repository = getCustomRepository(<%- entityCapitalize %>Repository);
      const <%- entityLowercase %>s = await repository.list(req.query);
      res.json(<%- entityLowercase %>s);
    }
    catch (e) { next(e); }
  }
<% } %>

<% if (options.delete) {%>
  /**
   * Delete <%- entityLowercase %>
   *
   * @param req Request
   * @param res Response
   * @param next Function
   *
   * @public
   */
  public async remove (req: Request, res : Response, next: Function) {

    try {
      const <%- entityLowercase %>Repository = getRepository(<%- entityCapitalize %>);
      const <%- entityLowercase %> = await <%- entityLowercase %>Repository.findOne(req.params.<%- entityLowercase %>Id);
      const repository = getRepository(<%- entityCapitalize %>);
      await repository.remove(<%- entityLowercase %>);
      res.sendStatus(HttpStatus.NO_CONTENT).end();
    }
    catch(e) { next(e); }

  }
<% } %>
}
