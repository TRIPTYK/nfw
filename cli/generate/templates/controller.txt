import * as HttpStatus from "http-status";

import { Request, Response } from "express";
import { {{ENTITY_CAPITALIZE}} } from "./../models/{{ENTITY_LOWERCASE}}.model";
import { {{ENTITY_CAPITALIZE}}Repository } from "./../repositories/{{ENTITY_LOWERCASE}}.repository";
import { getRepository, getCustomRepository } from "typeorm";
import { BaseController } from "./base.controller";
const promisify = require('util')

/**
 *
 */
export class {{ENTITY_CAPITALIZE}}Controller extends BaseController {

  /** */
  constructor() { super(); }

{{ENTITY_CRUD_READ_ID_START}}
  /**
   * Get serialized {{ENTITY_LOWERCASE}}
   *
   * @param req Request
   * @param res Response
   *
   * @public
   */
   public async get(req: Request, res : Response) {
    const repository = getRepository({{ENTITY_CAPITALIZE}});
    const {{ENTITY_LOWERCASE}} = await repository.findOneOrFail(req.params.{{ENTITY_LOWERCASE}}Id);
    res.json( {{ENTITY_LOWERCASE}}.whitelist() );
  }
{{ENTITY_CRUD_READ_ID_END}}

{{ENTITY_CRUD_CREATE_START}}
  /**
   * Create new {{ENTITY_LOWERCASE}}
   *
   * @param req Request
   * @param res Response
   * @param next Function
   *
   * @public
   */
  public async create (req: Request, res : Response, next: Function) {
    try {
      const repository = getRepository({{ENTITY_CAPITALIZE}});
      const {{ENTITY_LOWERCASE}} = new {{ENTITY_CAPITALIZE}}(req.body);
      const saved{{ENTITY_CAPITALIZE}} = await repository.save({{ENTITY_LOWERCASE}});
      res.status( HttpStatus.CREATED );
      res.json( saved{{ENTITY_CAPITALIZE}}.whitelist() );
    }
    catch (e) { next( e.message ); }
  }{{ENTITY_CRUD_CREATE_END}}

{{ENTITY_CRUD_UPDATE_PUT_START}}
  /**
   * Update existing {{ENTITY_LOWERCASE}}
   *
   * @param req Request
   * @param res Response
   * @param next Function
   *
   * @public
   */
  public async update (req: Request, res : Response, next: Function) {

    try {
      const repository = getRepository({{ENTITY_CAPITALIZE}});
      const {{ENTITY_LOWERCASE}} = await repository.findOne(req.params.{{ENTITY_LOWERCASE}}Id);
      repository.merge({{ENTITY_LOWERCASE}}, req.body);
      repository.save({{ENTITY_LOWERCASE}});
      res.json( {{ENTITY_LOWERCASE}}.whitelist() );
    }
    catch(e) { next( e.message ); }

  };
{{ENTITY_CRUD_UPDATE_PUT_END}}

{{ENTITY_CRUD_READ_START}}
  /**
   * Get {{ENTITY_LOWERCASE}} list
   *
   * @param req Request
   * @param res Response
   * @param next Function
   *
   * @public
   */
  public async list (req: Request, res : Response, next: Function) {

    try {
      const repository = getCustomRepository({{ENTITY_CAPITALIZE}}Repository);
      const {{ENTITY_LOWERCASE}}s = await repository.list(req.query);
      res.json({{ENTITY_LOWERCASE}}s);
    }
    catch (e) { next(e); }
  }
{{ENTITY_CRUD_READ_END}}

{{ENTITY_CRUD_DELETE_START}}
  /**
   * Delete {{ENTITY_LOWERCASE}}
   *
   * @param req Request
   * @param res Response
   * @param next Function
   *
   * @public
   */
  public async remove (req: Request, res : Response, next: Function) {

    try {
      const {{ENTITY_LOWERCASE}}Repository = getRepository({{ENTITY_CAPITALIZE}});
      const {{ENTITY_LOWERCASE}} = await {{ENTITY_LOWERCASE}}Repository.findOne(req.params.{{ENTITY_LOWERCASE}}Id);
      const repository = getRepository({{ENTITY_CAPITALIZE}});
      await repository.remove({{ENTITY_LOWERCASE}});
      res.sendStatus(HttpStatus.NO_CONTENT).end();
    }
    catch(e) { next(e); }

  }
{{ENTITY_CRUD_DELETE_END}}
}
