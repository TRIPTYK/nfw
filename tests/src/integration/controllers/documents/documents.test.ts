import { MikroORM } from '@mikro-orm/core';
import { container } from '@triptyk/nfw-core';
import { afterAll, beforeAll, expect, vi } from 'vitest';
import { DocumentsController } from 'app/api/controllers/documents.controller.js';
import { Roles } from 'app/api/enums/roles.enum.js';
import { UserModel } from 'app/database/models/user.model.js';
import { testCtx } from '../../../../utils/it-request-context.js';
import { setupIntegrationTest } from '../../../../utils/setup-integration-test.js';
import { deleteDummyDocument, DocumentsControllerTestSeeder, dummyDocument } from './seed.js';
import { unlink } from 'fs/promises';
import { MimeTypes } from 'app/api/enums/mime-type.enum.js';

let documentsController: DocumentsController;

beforeAll(async () => {
  await setupIntegrationTest(DocumentsControllerTestSeeder);
  documentsController = container.resolve(DocumentsController);
})

function createAdminUser () {
  const adminUser = new UserModel();
  adminUser.role = Roles.ADMIN;
  return adminUser;
}

testCtx('GetOne', () => container.resolve(MikroORM), async () => {
  const document = await documentsController.get('document', {}, new UserModel());

  expect(document).toMatchSnapshot();
});

testCtx('GetAll', () => container.resolve(MikroORM), async () => {
  const document = await documentsController.findAll({}, new UserModel());

  expect(document).toMatchSnapshot();
});

testCtx('CreateOne', () => container.resolve(MikroORM), async () => {
  const document = {
    filename: 'create-file.bmp',
    originalName: 'create-original-name.bmp',
    mimetype: MimeTypes.BMP,
    size: 1337,
    path: 'create-file.bmp',
    users: []
  };

  const constrollerResponse = await documentsController.create(document, createAdminUser());

  expect(constrollerResponse).toMatchObject({
    data: {
      attributes: {
        filename: 'create-file.bmp',
        mimetype: 'image/bmp',
        originalName: 'create-original-name.bmp',
        path: 'create-file.bmp',
        size: 1337,
      },
      relationships: undefined,
      type: 'documents',
    },
    included: undefined,
  });
});

testCtx('Update', () => container.resolve(MikroORM), async () => {
  const updatedDummyDocument = {
    ...dummyDocument,
    users: [],
    filename: 'update-filename',
  };
  const user = await documentsController.update('document', updatedDummyDocument, createAdminUser());

  expect(user).toMatchSnapshot();
});

vi.mock('fs/promises', async () => {
  // eslint-disable-next-line @typescript-eslint/consistent-type-imports
  const actualImport = await vi.importActual<typeof import('fs/promises')>('fs/promises');
  return {
    ...actualImport,
    unlink: vi.fn()
  }
})

testCtx('Delete', () => container.resolve(MikroORM), async () => {
  const user = await documentsController.delete('delete-document', createAdminUser());
  expect(unlink).toBeCalledWith(deleteDummyDocument.path);
  expect(user).toStrictEqual(null);
});

afterAll(() => {
  vi.restoreAllMocks();
})
