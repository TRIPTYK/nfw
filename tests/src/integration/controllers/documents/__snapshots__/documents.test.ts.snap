// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GetAll 1`] = `
{
  "data": [
    {
      "attributes": {
        "filename": "delete-filename.bmp",
        "mimetype": "image/bmp",
        "originalName": "delete-original-name.bmp",
        "path": "./delete-filename.bmp",
        "size": 1337,
      },
      "id": "delete-document",
      "links": {
        "self": "/api/v1/documents/delete-document",
      },
      "meta": undefined,
      "relationships": undefined,
      "type": "documents",
    },
    {
      "attributes": {
        "filename": "filename.bmp",
        "mimetype": "image/bmp",
        "originalName": "originalName.bmp",
        "path": "filename.bmp",
        "size": 8001,
      },
      "id": "document",
      "links": {
        "self": "/api/v1/documents/document",
      },
      "meta": undefined,
      "relationships": undefined,
      "type": "documents",
    },
  ],
  "included": undefined,
  "jsonapi": {
    "version": "1.0",
  },
  "links": {
    "self": "/api/v1/documents",
  },
  "meta": undefined,
}
`;

exports[`GetOne 1`] = `
{
  "data": {
    "attributes": {
      "filename": "filename.bmp",
      "mimetype": "image/bmp",
      "originalName": "originalName.bmp",
      "path": "filename.bmp",
      "size": 8001,
    },
    "id": "document",
    "links": {
      "self": "/api/v1/documents/document",
    },
    "meta": undefined,
    "relationships": undefined,
    "type": "documents",
  },
  "included": undefined,
  "jsonapi": {
    "version": "1.0",
  },
  "links": {
    "self": "/api/v1/documents/document",
  },
  "meta": undefined,
}
`;

exports[`Update 1`] = `
{
  "data": {
    "attributes": {
      "filename": "update-filename",
      "mimetype": "image/bmp",
      "originalName": "originalName.bmp",
      "path": "filename.bmp",
      "size": 8001,
    },
    "id": "document",
    "links": {
      "self": "/api/v1/documents/document",
    },
    "meta": undefined,
    "relationships": undefined,
    "type": "documents",
  },
  "included": undefined,
  "jsonapi": {
    "version": "1.0",
  },
  "links": {
    "self": "/api/v1/documents/document",
  },
  "meta": undefined,
}
`;
